src/python_testing/TC_DGWIFI_2_2.py

```diff
@@ -0,0 +1,219 @@
+        )
+
+    #
+    # Methods to cause/triggers in your environment
+    #
+    async def send_wifi_disconnection_test_event_trigger(self):
+        # Send test event trigger to programmatically cause a Wi-Fi disconnection in the DUT.
+        await self.send_test_event_triggers(eventTrigger=0x0036000000000000)
+
+    async def read_disconnection_events(self, endpoint):
+        event_path = [(endpoint, Clusters.WiFiNetworkDiagnostics.Events.Disconnection, 0)]
+        return await self.default_controller.ReadEvent(
+            nodeid=self.dut_node_id, events=event_path
+        )
+
+    async def send_wifi_association_failure_test_event_trigger(self):
+        # Send test event trigger to programmatically cause repeated association failures on the DUT.
+        await self.send_test_event_triggers(eventTrigger=0x0036000000000001)
+
+    async def read_association_failure_events(self, endpoint):
+        event_path = [(endpoint, Clusters.WiFiNetworkDiagnostics.Events.AssociationFailure, 0)]
+        return await self.default_controller.ReadEvent(
+            nodeid=self.dut_node_id, events=event_path
+        )
+
+    async def send_wifi_reconnection_test_event_trigger(self):
+        # Send test event trigger to programmatically reconnect the DUT to Wi-Fi.
+        await self.send_test_event_triggers(eventTrigger=0x0036000000000002)
+
+    async def read_connection_status_events(self, endpoint):
+        event_path = [(endpoint, Clusters.WiFiNetworkDiagnostics.Events.ConnectionStatus, 0)]
+        return await self.default_controller.ReadEvent(
+            nodeid=self.dut_node_id, events=event_path
+        )
+
+    #
+    # Test description & PICS
+    #
+    def desc_TC_DGWIFI_2_2(self) -> str:
+        """Returns a description of this test"""
+        return "[TC-DGWIFI-2.2] Wi-Fi Diagnostics Event Tests (Server as DUT)"
+
+    def pics_TC_DGWIFI_2_2(self) -> list[str]:
+        return ["DGWIFI.S"]
+
+    def steps_TC_DGWIFI_2_2(self) -> list[TestStep]:
+        steps = [
+            TestStep(1, "Commissioning (already done)", is_commissioning=True),
+            TestStep(2, "Trigger Wi-Fi disconnection -> verify Disconnection event"),
+            TestStep(3, "Trigger repeated association failures -> verify AssociationFailure event"),
+            TestStep(4, "Reconnect Wi-Fi -> verify ConnectionStatus event"),
+        ]
+        return steps
+
+    # ---------------------------
+    # Main Test Routine
+    # ---------------------------
+    @async_test_body
+    async def test_TC_DGWIFI_2_2(self):
+        endpoint = self.get_endpoint(default=0)
+
+        #
+        # STEP 1: Commission DUT (already done)
+        #
+        self.step(1)
+
+        #
+        # STEP 2: Cause a Wi-Fi Disconnection -> read Disconnection event from the DUT
+        #
+        self.step(2)
+        await self.send_wifi_disconnection_test_event_trigger()
+
+        # Allow time for the event to propagate
+        await asyncio.sleep(1)
```

Comment:
is there a way for `wait with timeout` instead of a hardcoded sleep? I would be nice to have a logic of "will receive event within 5 seconds" for example instead of sleep and hope we get something.

Reply:
Yes, we can do that